src/runtime/naming/name.cpp

```diff
@@ -421,20 +422,31 @@ namespace hpx { namespace naming
         ///////////////////////////////////////////////////////////////////////
         std::int64_t replenish_credits(gid_type& gid)
         {
+            typedef std::unique_lock<gid_type::mutex_type> scoped_lock;
+            scoped_lock l(gid);
+            return replenish_credits_locked(l, gid);
+        }
+
+        HPX_EXPORT std::int64_t replenish_credits_locked(
+            std::unique_lock<gid_type::mutex_type>& l, gid_type& gid)
+        {
             std::int64_t added_credit = 0;
 
-            {
-                typedef std::unique_lock<gid_type::mutex_type> scoped_lock;
-                scoped_lock l(gid);
-                HPX_ASSERT(0 == get_credit_from_gid(gid));
+            HPX_ASSERT(0 == get_credit_from_gid(gid));
 
-                added_credit = naming::detail::fill_credit_for_gid(gid);
-                naming::detail::set_credit_split_mask_for_gid(gid);
-            }
+            added_credit = naming::detail::fill_credit_for_gid(gid);
+            naming::detail::set_credit_split_mask_for_gid(gid);
 
             gid_type unlocked_gid = gid;        // strips lock-bit
 
-            return agas::incref(unlocked_gid, added_credit);
+            std::int64_t result = 0;
+            {
+                hpx::util::scoped_unlock<std::unique_lock<gid_type::mutex_type> >
+                    ul(l);
+                 result = agas::incref(unlocked_gid, added_credit);
+            }
+
+            return result;
```

Comment:
> One problem I discovered was that the promise gid never had credits. I changed it such that it replenishes the credit at the first call. We need to hold a lock on the gid while checking the credit count and replenishing, that's why I introduced that function.

That was intentional. The first batch of credits was given away to the first caller of `get_id()` to avoid credit splitting in the common case. Credits were supposed to be replenished (and stored) only starting with the second call to `get_id()`.

Your fix does not add holding a lock while replenishing the credit. That was in place before. So I still don't understand the rationale of the change you're proposing.